compiler/rustc_hir/src/def.rs

```diff
@@ -314,68 +314,75 @@ pub enum Res<Id = hir::HirId> {
     /// **Belongs to the type namespace.**
     PrimTy(hir::PrimTy),
 
-    /// The `Self` type, optionally with the [`DefId`] of the trait it belongs to and
-    /// optionally with the [`DefId`] of the item introducing the `Self` type alias.
+    /// The `Self` type, as used within a trait.
     ///
     /// **Belongs to the type namespace.**
     ///
-    /// Examples:
+    /// Example:
     /// ```
-    /// struct Bar(Box<Self>);
-    /// // `Res::SelfTy { trait_: None, alias_of: Some(Bar) }`
-    ///
     /// trait Foo {
-    ///     fn foo() -> Box<Self>;
-    ///     // `Res::SelfTy { trait_: Some(Foo), alias_of: None }`
+    ///     fn foo() -> Box<Self>; // `Res::SelfTyTrait { trait_: Foo }`
     /// }
+    /// ```
+    SelfTyTrait {
+        /// The trait this `Self` is a generic arg for.
+        trait_: DefId,
+    },
+
+    /// The `Self` type, as used somewhere other than within a trait.
+    ///
+    /// **Belongs to the type namespace.**
+    ///
+    /// Examples:
+    /// ```
+    /// struct Bar(Box<Self>);
     ///
     /// impl Bar {
     ///     fn blah() {
     ///         let _: Self;
-    ///         // `Res::SelfTy { trait_: None, alias_of: Some(::{impl#0}) }`
     ///     }
     /// }
     ///
     /// impl Foo for Bar {
-    ///     fn foo() -> Box<Self> {
-    ///     // `Res::SelfTy { trait_: Some(Foo), alias_of: Some(::{impl#1}) }`
+    ///     fn foo() -> Box<Self>
     ///         let _: Self;
-    ///         // `Res::SelfTy { trait_: Some(Foo), alias_of: Some(::{impl#1}) }`
     ///
     ///         todo!()
     ///     }
     /// }
     /// ```
-    ///
     /// *See also [`Res::SelfCtor`].*
     ///
-    /// -----
-    ///
-    /// HACK(min_const_generics): self types also have an optional requirement to **not** mention
-    /// any generic parameters to allow the following with `min_const_generics`:
-    /// ```
-    /// # struct Foo;
-    /// impl Foo { fn test() -> [u8; std::mem::size_of::<Self>()] { todo!() } }
-    ///
-    /// struct Bar([u8; baz::<Self>()]);
-    /// const fn baz<T>() -> usize { 10 }
-    /// ```
-    /// We do however allow `Self` in repeat expression even if it is generic to not break code
-    /// which already works on stable while causing the `const_evaluatable_unchecked` future compat
-    /// lint:
-    /// ```
-    /// fn foo<T>() {
-    ///     let _bar = [1_u8; std::mem::size_of::<*mut T>()];
-    /// }
-    /// ```
-    // FIXME(generic_const_exprs): Remove this bodge once that feature is stable.
-    SelfTy {
-        /// The trait this `Self` is a generic arg for.
-        trait_: Option<DefId>,
+    SelfTyOther {
```

Comment:
```suggestion
    SelfTyAlias {
```
For consistency with `DefKind::TyAlias`, of which this variant is an equivalent.

Reply:
The comment on `DefKind::TyAlias` says:
```rust
    /// Type alias: `type Foo = Bar;`
```
Again, AFAICT this is entirely different to the `Self` type within a `struct`/`impl` that `Res::SelfTyOther` represents.

Reply:
Semantically `SelfTypeAlias` is exactly the same thing as `DefKind::TyAlias`, the only difference is that it's introduced implicitly rather than through explicit `type` syntax.